### PR TITLE
API updates

### PR DESCRIPTION
**Pull request by tulimaki**
This PR adds and amends APIs for storefront usage.

* new APIs were added for addresses, contacts, users, order history, and basket. 
* Product APIs have been updated significantly for improved usability
* the shop API was updated to support filtering by distance
* Order creation was amended to allow orders to be created without a customer, addresses, or service methods
* Database indices were added to order/product/shops `created_on/modified_on` for client caching purposes
* Most basket logic was moved from front to core so that storefronts can be built without needing `shuup.front`

Replaces https://github.com/shuup/shuup/pull/1063

### CODE REVIEW

File: shuup/admin/modules/orders/json_order_creator.py [hunk 1]
```diff
@@ -157,7 +162,7 @@ def _process_lines(self, source, state):
             try:
                 self._add_json_line_to_source(source, sline)
             except Exception as exc:  # pragma: no cover
-                self.add_error(exc)
+                self.add_error(ValidationError(str(exc), code="line_error"))
```
Comment (suutari-ai) [inline shuup/admin/modules/orders/json_order_creator.py]: Validation error should be (a) unicode (on py2) and (b) translated (user visible) message.

File: shuup/core/api/front_products.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+        elif image.file:
+            return self.context["request"].build_absolute_uri(image.file.url)
+
+    def get_attributes(self, shop_product):
+        return AttributeSerializer(shop_product.product.attributes, many=True).data
+
+    def get_price(self, shop_product):
+        key, val = context_cache.get_cached_value(identifier="shop_product_price",
+                                                  item=shop_product,
+                                                  context={},
```
Comment (tulimaki) [inline shuup/core/api/front_products.py]: Fix context here.

File: shuup/core/models/_shops.py [hunk 3]
```diff
@@ -43,7 +43,7 @@ class Shop(ChangeProtected, TranslatableShuupModel):
     change_protect_message = _("The following fields cannot be changed since there are existing orders for this shop")
```
Comment (tulimaki) [inline shuup/core/models/_shops.py]: Should shop status, longitude, latitude has also `db_index`?

File: setup.cfg [hunk 4]
```diff
@@ -17,7 +17,7 @@ combine_as_imports=false
 indent=4
 known_first_party=shuup
 known_standard_library=token,tokenize,enum,importlib
-known_third_party=django,six
+known_third_party=django,six,rest_framework,django_filters
```
Comment (tulimaki) [inline setup.cfg]: Is this really needed?

**Comment by tulimaki** [comment]
As long as we don't want to squash this. I think we don't.

Fix following commit messages:
`Basket fixes`
`It's possible to create basket for certain customer`
`Add basket changes`

**Pull request merged**